dialogue.go

```diff
@@ -190,26 +193,29 @@ func (d *Dialogue) MarshalLen() int {
 	if field := d.ObjectIdentifier; field != nil {
 		l += field.MarshalLen()
 	}
-	if field := d.SingleAsn1Type; field != nil {
-		l += field.MarshalLen()
-	}
 	if field := d.DialoguePDU; field != nil {
-		l += field.MarshalLen()
+		l += field.MarshalLen() + 2 // 2 = singleAsn1Type IE Header
 	}
-	l += len(d.Payload)
 
-	return l
+	return l + len(d.Payload)
 }
 
 // SetLength sets the length in Length field.
 func (d *Dialogue) SetLength() {
+	if field := d.ObjectIdentifier; field != nil {
+		d.ObjectIdentifier.SetLength()
+	}
+	if field := d.DialoguePDU; field != nil {
+		d.DialoguePDU.SetLength()
+	}
+
 	d.Length = uint8(d.MarshalLen() - 2)
 	d.ExternalLength = uint8(d.MarshalLen() - 4)
 }
 
 // String returns the SCCP common header values in human readable format.
 func (d *Dialogue) String() string {
-	return fmt.Sprintf("{Tag: %#x, Length: %d, ExternalTag: %x, ExternalLength: %d, ObjectIdentifier: %v, SingleAsn1Type: %v, DialoguePDU: %v, Payload: %x}",
+	return fmt.Sprintf("{Tag: %#x, Length: %d, ExternalTag: %x, ExternalLength: %d, ObjectIdentifier: %v, singleAsn1Type: %v, DialoguePDU: %v, Payload: %x}",
```

Comment:
For the consistency (exact field name should be used)

```suggestion
	return fmt.Sprintf("{Tag: %#x, Length: %d, ExternalTag: %x, ExternalLength: %d, ObjectIdentifier: %v, SingleAsn1Type: %v, DialoguePDU: %v, Payload: %x}",
```